## Stock Data
Now that you've had exposure to time series data, let's look at bringing stock prices into Pandas.
## Reading in Data
Your dataset can come in a variety of different formats. The most common format is the [CSV](https://en.wikipedia.org/wiki/Comma-separated_values). We'll use the "prices.csv" file as an example csv file.

In [1]:
with open('prices.csv', 'r') as file:
    prices = file.read()
    
print(prices)

ABC,2017-09-05,163.75,164.25,160.56,162.08,29317054.0,162.08,29317054.0
ABC,2017-09-06,162.71,162.99,160.52,161.91,21179047.0,161.91,21179047.0
ABC,2017-09-07,162.09,162.24,160.36,161.26,21722995.0,161.26,21722995.0
ABC,2017-09-08,160.86,161.15,158.53,158.63,28183159.0,158.63,28183159.0
ABC,2017-09-11,160.5,162.05,159.89,161.5,31028926.0,161.5,31028926.0
ABC,2017-09-12,162.61,163.96,158.77,160.82,71139119.0,160.82,71139119.0
ABC,2017-09-13,159.87,159.96,157.91,159.65,44393752.0,159.65,44393752.0
EFG,2017-09-05,154.83,155.41,153.31,154.28,1268901.0,154.28,1268901.0
EFG,2017-09-06,155.11,155.555,153.61,153.76,1200200.0,153.76,1200200.0
EFG,2017-09-07,154.49,155.885,153.64,155.47,1414048.0,155.47,1414048.0
EFG,2017-09-08,155.24,155.88,154.29,155.34,1441080.0,155.34,1441080.0
EFG,2017-09-11,156.59,157.89,156.38,156.87,1612915.0,156.87,1612915.0
EFG,2017-09-12,156.76,157.2152,154.92,156.31,1696179.0,156.31,1696179.0
EFG,2017-09-13,155.88,156.7,155.39,156.24,1212593.0,156.24,1212593.0
XYZ,20

The data provider will provide you with information for each field in the CSV. This csv has the fields ticker, date, open, high, low, close, volume, adj_close, adj_volume in that order. That means, the first line in the CSV has the following data:

- ticker: ABC
- date: 2017-09-05
- open: 163.75
- high: 164.25
- low: 160.56
- close: 162.08
- volume: 29317054.0
- adj_close: 162.08
- adj_volume: 29317054.0

Let's move this data into a DataFrame. For this, we'll need to use the [`pd.read_csv`](https://pandas.pydata.org/pandas-docs/version/0.21.1/generated/pandas.read_csv.html) function. This allows you generate a DataFrame from CSV data.

In [2]:
import pandas as pd

price_df = pd.read_csv('prices.csv')

price_df

ABC  2017-09-05  163.75    164.25  160.56  162.08  29317054.0    162.08.1  \
0   ABC  2017-09-06  162.71  162.9900  160.52  161.91  21179047.0  161.910000   
1   ABC  2017-09-07  162.09  162.2400  160.36  161.26  21722995.0  161.260000   
2   ABC  2017-09-08  160.86  161.1500  158.53  158.63  28183159.0  158.630000   
3   ABC  2017-09-11  160.50  162.0500  159.89  161.50  31028926.0  161.500000   
4   ABC  2017-09-12  162.61  163.9600  158.77  160.82  71139119.0  160.820000   
5   ABC  2017-09-13  159.87  159.9600  157.91  159.65  44393752.0  159.650000   
6   EFG  2017-09-05  154.83  155.4100  153.31  154.28   1268901.0  154.280000   
7   EFG  2017-09-06  155.11  155.5550  153.61  153.76   1200200.0  153.760000   
8   EFG  2017-09-07  154.49  155.8850  153.64  155.47   1414048.0  155.470000   
9   EFG  2017-09-08  155.24  155.8800  154.29  155.34   1441080.0  155.340000   
10  EFG  2017-09-11  156.59  157.8900  156.38  156.87   1612915.0  156.870000   
11  EFG  2017-09-12  156.76  157.2152  154.92  156.31   1696179.0  156.310000   
12  EFG  2017-09-13  155.88  156.7000  155.39  156.24   1212593.0  156.240000   
13  XYZ  2017-09-05   64.10   64.5100   63.31   63.64   1735223.0   63.451220   
14  XYZ  2017-09-06   63.82   63.8200   61.84   61.98   3725588.0   61.796144   
15  XYZ  2017-09-07   61.97   61.9900   59.25   60.50   6171796.0   60.500000   
16  XYZ  2017-09-08   60.24   60.3900   58.68   59.42   5195595.0   59.420000   
17  XYZ  2017-09-11   59.79   59.9200   57.92   58.00   4980384.0   58.000000   
18  XYZ  2017-09-12   58.06   59.4500   57.63   58.82   3633834.0   58.820000   
19  XYZ  2017-09-13   59.00   60.6000   59.00   60.51   3588591.0   60.510000   

    29317054.0.1  
0     21179047.0  
1     21722995.0  
2     28183159.0  
3     31028926.0  
4     71139119.0  
5     44393752.0  
6      1268901.0  
7      1200200.0  
8      1414048.0  
9      1441080.0  
10     1612915.0  
11     1696179.0  
12     1212593.0  
13     1735223.0  
14     3725588.0  
15     6171796.0  
16     5195595.0  
17     4980384.0  
18     3633834.0  
19     3588591.0

That generated a DataFrame using the CSV, but assumed the first row contains the field names. We'll have to supply the function's parameter `names` with a list of fiels names.

In [3]:
price_df = pd.read_csv('prices.csv', names=['ticker', 'date', 'open', 'high', 'low',
                                             'close', 'volume', 'adj_close', 'adj_volume'])

price_df

ticker        date    open      high     low   close      volume  \
0     ABC  2017-09-05  163.75  164.2500  160.56  162.08  29317054.0   
1     ABC  2017-09-06  162.71  162.9900  160.52  161.91  21179047.0   
2     ABC  2017-09-07  162.09  162.2400  160.36  161.26  21722995.0   
3     ABC  2017-09-08  160.86  161.1500  158.53  158.63  28183159.0   
4     ABC  2017-09-11  160.50  162.0500  159.89  161.50  31028926.0   
5     ABC  2017-09-12  162.61  163.9600  158.77  160.82  71139119.0   
6     ABC  2017-09-13  159.87  159.9600  157.91  159.65  44393752.0   
7     EFG  2017-09-05  154.83  155.4100  153.31  154.28   1268901.0   
8     EFG  2017-09-06  155.11  155.5550  153.61  153.76   1200200.0   
9     EFG  2017-09-07  154.49  155.8850  153.64  155.47   1414048.0   
10    EFG  2017-09-08  155.24  155.8800  154.29  155.34   1441080.0   
11    EFG  2017-09-11  156.59  157.8900  156.38  156.87   1612915.0   
12    EFG  2017-09-12  156.76  157.2152  154.92  156.31   1696179.0   
13    EFG  2017-09-13  155.88  156.7000  155.39  156.24   1212593.0   
14    XYZ  2017-09-05   64.10   64.5100   63.31   63.64   1735223.0   
15    XYZ  2017-09-06   63.82   63.8200   61.84   61.98   3725588.0   
16    XYZ  2017-09-07   61.97   61.9900   59.25   60.50   6171796.0   
17    XYZ  2017-09-08   60.24   60.3900   58.68   59.42   5195595.0   
18    XYZ  2017-09-11   59.79   59.9200   57.92   58.00   4980384.0   
19    XYZ  2017-09-12   58.06   59.4500   57.63   58.82   3633834.0   
20    XYZ  2017-09-13   59.00   60.6000   59.00   60.51   3588591.0   

     adj_close  adj_volume  
0   162.080000  29317054.0  
1   161.910000  21179047.0  
2   161.260000  21722995.0  
3   158.630000  28183159.0  
4   161.500000  31028926.0  
5   160.820000  71139119.0  
6   159.650000  44393752.0  
7   154.280000   1268901.0  
8   153.760000   1200200.0  
9   155.470000   1414048.0  
10  155.340000   1441080.0  
11  156.870000   1612915.0  
12  156.310000   1696179.0  
13  156.240000   1212593.0  
14   63.451220   1735223.0  
15   61.796144   3725588.0  
16   60.500000   6171796.0  
17   59.420000   5195595.0  
18   58.000000   4980384.0  
19   58.820000   3633834.0  
20   60.510000   3588591.0

## DataFrame Calculations
Now that we have the data in a DataFrame, we can start to do calculations on it. Let's find out the median value for each stock using the [`DataFrame.medain`](https://pandas.pydata.org/pandas-docs/version/0.21/generated/pandas.DataFrame.median.html) function.

In [4]:
price_df.median()

open              155.240
high              155.885
low               154.290
close             155.470
volume        3725588.000
adj_close         155.470
adj_volume    3725588.000
dtype: float64

That's not right. Those are the median values for the whole stock universe. We'll use the [`DataFrame.groupby`](https://pandas.pydata.org/pandas-docs/version/0.21/generated/pandas.DataFrame.groupby.html) function to get mean for each stock.

In [5]:
price_df.groupby('ticker').median()

open     high     low   close      volume  adj_close  adj_volume
ticker                                                                    
ABC     162.09  162.240  159.89  161.26  29317054.0     161.26  29317054.0
EFG     155.24  155.885  154.29  155.47   1414048.0     155.47   1414048.0
XYZ      60.24   60.600   59.00   60.50   3725588.0      60.50   3725588.0

That's what we're looking for! However, we don't want to run the `groupby` function each time we make an operation. We could save the GroupBy object by doing `price_df_ticker_groups = price_df.groupby('ticker')`. This limits us to the operations of GroupBy objects. There's the [`GroupBy.apply`](https://pandas.pydata.org/pandas-docs/version/0.21/generated/pandas.core.groupby.GroupBy.apply.html), but then we lose out on performance. The true problem is the way the data is represented.

In [6]:
price_df.iloc[:10]

ticker        date    open     high     low   close      volume  adj_close  \
0    ABC  2017-09-05  163.75  164.250  160.56  162.08  29317054.0     162.08   
1    ABC  2017-09-06  162.71  162.990  160.52  161.91  21179047.0     161.91   
2    ABC  2017-09-07  162.09  162.240  160.36  161.26  21722995.0     161.26   
3    ABC  2017-09-08  160.86  161.150  158.53  158.63  28183159.0     158.63   
4    ABC  2017-09-11  160.50  162.050  159.89  161.50  31028926.0     161.50   
5    ABC  2017-09-12  162.61  163.960  158.77  160.82  71139119.0     160.82   
6    ABC  2017-09-13  159.87  159.960  157.91  159.65  44393752.0     159.65   
7    EFG  2017-09-05  154.83  155.410  153.31  154.28   1268901.0     154.28   
8    EFG  2017-09-06  155.11  155.555  153.61  153.76   1200200.0     153.76   
9    EFG  2017-09-07  154.49  155.885  153.64  155.47   1414048.0     155.47   

   adj_volume  
0  29317054.0  
1  21179047.0  
2  21722995.0  
3  28183159.0  
4  31028926.0  
5  71139119.0  
6  44393752.0  
7   1268901.0  
8   1200200.0  
9   1414048.0

Can you spot our problem? Take a moment to see if you can find it.

The problem is between lines [6, 7] and  [13, 14]

In [7]:
price_df.iloc[[6, 7, 13, 14]]

ticker        date    open    high     low   close      volume  adj_close  \
6     ABC  2017-09-13  159.87  159.96  157.91  159.65  44393752.0  159.65000   
7     EFG  2017-09-05  154.83  155.41  153.31  154.28   1268901.0  154.28000   
13    EFG  2017-09-13  155.88  156.70  155.39  156.24   1212593.0  156.24000   
14    XYZ  2017-09-05   64.10   64.51   63.31   63.64   1735223.0   63.45122   

    adj_volume  
6   44393752.0  
7    1268901.0  
13   1212593.0  
14   1735223.0

Data for all the tickers are stacked. We're representing 3 dimensional data in 2 dimensions. This was solved using Panda's Panels, which is [depricated](https://pandas.pydata.org/pandas-docs/version/0.21.0/dsintro.html#deprecate-panel). The Pandas documentation reccommends we use either [MultiIndex](https://pandas.pydata.org/pandas-docs/version/0.21/advanced.html) or [xarray](http://xarray.pydata.org/en/stable/). [MultiIndex](https://pandas.pydata.org/pandas-docs/version/0.21/advanced.html) still doesn't solve our problem, since the data is still represented in 2 dimensions. [xarray](http://xarray.pydata.org/en/stable/) is able to store 3 dimensional data, but Finance uses Pandas, so we'll stick with this library. After you finish this program, I reccommend you check out [xarray](http://xarray.pydata.org/en/stable/).

So, how do we use our 3-dimensional data with Pandas? We can split each 3rd dimension into it's own 2 dimension DataFrame. Let's take this array as an example:
```
[
    [
        [ 0,  1],
        [ 2,  3],
        [ 4,  5]
    ],[
        [ 6,  7],
        [ 8,  9],
        [10, 11]
    ],[
        [12, 13],
        [14, 15],
        [16, 17]
    ],[
        [18, 19],
        [20, 21],
        [22, 23]
    ]
]    
```
We want to split it into these two 2d arrays:
```
[
    [0, 2, 4],
    [6, 8, 10],
    [12, 14, 16],
    [18, 20, 22]
] 
[
    [1, 3, 5],
    [7, 8, 11],
    [13, 15, 17],
    [19, 21, 23]
] 
```
In our case, our third dimensions are "open", "high", "low", "close", "volume", "adj_close", and  "adj_volume". We'll use the [`DataFrame.pivot`](https://pandas.pydata.org/pandas-docs/version/0.21/generated/pandas.DataFrame.pivot.html) function to generate these DataFrames.

In [ ]:
open_prices = price_df.pivot(index='date', columns='ticker', values='open')
high_prices = price_df.pivot(index='date', columns='ticker', values='high')
low_prices = price_df.pivot(index='date', columns='ticker', values='low')
close_prices = price_df.pivot(index='date', columns='ticker', values='close')
volume = price_df.pivot(index='date', columns='ticker', values='volume')
adj_close_prices = price_df.pivot(index='date', columns='ticker', values='adj_close')
adj_volume = price_df.pivot(index='date', columns='ticker', values='adj_volume')

open_prices

That gives you DataFrames for all the open, high low, etc.. Now, what we have been waiting for.. The mean for each ticker.

In [ ]:
open_prices.mean()

We can also get the mean for each date by doing a transpose.

In [ ]:
open_prices.T.mean()

It doesn't matter whether date is the index and tickers are the colums or the other way around. It's always a transpose away. Since we're going to do a lot of operations across dates, we will stick with date as the index and tickers as the colums throughtout this program.
## Quiz
Let's see if you can apply what you learned. Implment the `csv_to_pivot` function to take in a filepath, `csv_filename`, and output the close 2d array. You can assume the CSV file used by `csv_to_close` has the same field names as "prices.csv" and in the same order.

To help with your implemention of quizzes, we provide you with unit tests to test your function implemention. For this quiz, we'll be using the function `test_csv_to_close` in the `quiz_tests` module to test `csv_to_close`.

In [ ]:
import quiz_tests


def csv_to_close(csv_filepath, field_names):
    """Reads in data from a csv file and produces a DataFrame with close data.
    
    Parameters
    ----------
    csv_filepath : str
        The name of the csv file to read
    field_names : list of str
        The field names of the field in the csv file

    Returns
    -------
    close : DataFrame
        Close prices for each ticker and date
    """
    
    # TODO: Implement Function
    
    return None


quiz_tests.test_csv_to_close(csv_to_close)

## Quiz Solution
If you're having trouble, you can check out the quiz solution [here](stock_data_solution.ipynb).